### PR TITLE
✨ Add scale subresource logic to fake client

### PR DESCRIPTION
**Pull request by TheSpiritXIII**
Adds basic GET/UPDATE support to the scale sub-resource according to the docs: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/client#SubResourceClientConstructor

### CODE REVIEW

File: pkg/client/fake/client.go [hunk 1]
```diff
@@ -1278,3 +1315,88 @@ func zero(x interface{}) {
...
+	case *corev1.ReplicationController:
+		obj.Spec.Replicas = &scale.Spec.Replicas
+	case *appsv1.StatefulSet:
+		obj.Spec.Replicas = &scale.Spec.Replicas
+	default:
+		// TODO: CRDs https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource
+		return fmt.Errorf("unable to extract scale from type %T", obj)
+	}
+	return nil
+}
```
Comment (troy0820) [inline pkg/client/fake/client.go:1402]: Wouldn't this not allow people who make CRs who wish to use the sub resource scale be limited to only allow the types in the case statement available?

File: pkg/client/fake/client.go [hunk 2]
```diff
@@ -1080,7 +1082,24 @@ type fakeSubResourceClient struct {
...
+		}
+		scaleOut, err := extractScale(obj)
+		if err != nil {
+			return err
+		}
+		*scale = scaleOut
+		return nil
+	default:
+		return fmt.Errorf("fakeSubResourceClient does not support get for %s", sw.subResource)
+	}
```
Comment (troy0820) [inline pkg/client/fake/client.go:1102]: Having to use the subresource client to then use the actual client seems like we wouldn't need the sub resource client to do the GET.  The value of the scale is indicative on the spec and the subresource client has to get that value from the actual client where it's derived.

File: pkg/client/fake/client.go [hunk 3]
```diff
@@ -1108,10 +1127,28 @@ func (sw *fakeSubResourceClient) Update(ctx context.Context, obj client.Object,
 	updateOptions.ApplyOptions(opts)
 
 	body := obj
-	if updateOptions.SubResourceBody != nil {
-		body = updateOptions.SubResourceBody
+	switch sw.subResource {
+	case "scale":
```
Comment (alvaroaleman) [inline pkg/client/fake/client.go:1131]: Lets use a `const` for the subresource name

File: pkg/client/fake/client.go [hunk 4]
```diff
@@ -1108,10 +1127,28 @@ func (sw *fakeSubResourceClient) Update(ctx context.Context, obj client.Object,
...
+		if updateOptions.SubResourceBody == nil {
+			return apierrors.NewBadRequest("expected SubResourceBody")
+		}
+		scale, isScale := updateOptions.SubResourceBody.(*autoscalingv1.Scale)
+		if !isScale {
+			return apierrors.NewBadRequest(fmt.Sprintf("got invalid type %t, expected Scale", updateOptions.SubResourceBody))
```
Comment (alvaroaleman) [inline pkg/client/fake/client.go:1137]: Is this the behavior of the KAS, to first complain with a bad request about the request body before checking if the main resource even exists?

File: pkg/client/fake/client.go [hunk 5]
```diff
@@ -1278,3 +1315,88 @@ func zero(x interface{}) {
 	res := reflect.ValueOf(x).Elem()
 	res.Set(reflect.Zero(res.Type()))
 }
+
+func extractScale(obj client.Object) (autoscalingv1.Scale, error) {
+	switch obj := obj.(type) {
+	case *appsv1.Deployment:
```
Comment (alvaroaleman) [inline pkg/client/fake/client.go:1321]: Is go smart enough to allow expression this in a single switch case:
```suggestion
	case *appsv1.Deployment, *appsv1.ReplicaSet:
```

File: pkg/client/fake/client.go [hunk 6]
```diff
@@ -1278,3 +1315,88 @@ func zero(x interface{}) {
...
+				Replicas: replicas,
+			},
+			Status: autoscalingv1.ScaleStatus{
+				Replicas: obj.Status.Replicas,
+				Selector: obj.Spec.Selector.String(),
+			},
+		}, nil
+	default:
+		// TODO: CRDs https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource
+		return autoscalingv1.Scale{}, fmt.Errorf("unable to extract scale from type %T", obj)
```
Comment (alvaroaleman) [inline pkg/client/fake/client.go:1383]: This is not a good error, the error should clarify that scale is not implemented in the fake client for %T. Ideally we would also exclude all in-tree resources that do not have a scale subresource and return the apprpropriate error for that rather than falling through here

**Review by alvaroaleman — COMMENTED** [review]
Thanks a lot for looking into this, really appreciated! Few comments, we need to make really sure we behave the same way as the KAS does everywhere, because as soon as we release this, some ppl will start depending on whatever behavior it exhibits which might cause tests to incorrectly pass if this doesn't exactly match the kube apiserver behavior.

File: pkg/client/fake/client.go (hunk 6 above)
Comment (TheSpiritXIII) [reply pkg/client/fake/client.go]: Thanks! Added a look-up and more details here. Please let me know if you'd like any further changes.

File: pkg/client/fake/client.go (hunk 5 above)
Comment (TheSpiritXIII) [reply pkg/client/fake/client.go:1326]: No, sadly :/

File: pkg/client/fake/client.go (hunk 2 above)
Comment (TheSpiritXIII) [reply pkg/client/fake/client.go]: This is implemented on the server-side as far as I know. I cannot just lookup on the client because the fake client doesn't know how to extract this information. It's something of a meta-resource.

File: pkg/client/fake/client.go (hunk 4 above)
Comment (TheSpiritXIII) [reply pkg/client/fake/client.go]: I was unable to find it; am I looking at the right place? https://github.com/search?q=repo%3Akubernetes%2Fapiserver+scale+-path%3Atest&type=code

Nonetheless, I was able to reproduce it with `client.Client.Get` and your intuition is correct: it looks up the object first! Great catch.

File: pkg/client/fake/client.go (hunk 2 above)
Comment (troy0820) [reply pkg/client/fake/client.go]: Fair and not criticizing the implementation but asking because you need to deal with this the way you are dealing with it.

**Comment by alvaroaleman** [comment]
Will try to give this a proper review this week, sorry for the delay.
/ok-to-test

File: pkg/client/fake/client.go (hunk 4 above)
Comment (alvaroaleman) [reply pkg/client/fake/client.go]: I think the upstream code is here: https://github.com/kubernetes/kubernetes/blob/fb6bbc9781d11a87688c398778525c4e1dcb0f08/pkg/registry/apps/deployment/storage/storage.go#L320

Not exactly a quick read though it seems.

File: pkg/client/fake/client.go [hunk 7]
```diff
@@ -1278,3 +1322,126 @@ func zero(x interface{}) {
...
+			Spec: autoscalingv1.ScaleSpec{
+				Replicas: replicas,
+			},
+			Status: autoscalingv1.ScaleStatus{
+				Replicas: obj.Status.Replicas,
+				Selector: obj.Spec.Selector.String(),
+			},
+		}, nil
+	default:
+		crd, err := getCustomResourceDefinition(ctx, c, obj)
```
Comment (alvaroaleman) [inline pkg/client/fake/client.go:1389]: Can we remove this code and directly error out? For status for example we have an explicit config option on the fake client. I don't know how that would best look like for scale given it requires more config than just a binary enable/disable, but I would prefer not adding any code related to that until we actually support that.

File: pkg/client/fake/client.go [hunk 8]
```diff
@@ -1278,3 +1322,126 @@ func zero(x interface{}) {
...
+
+func extractScale(ctx context.Context, c client.Client, obj client.Object) (*autoscalingv1.Scale, error) {
+	switch obj := obj.(type) {
+	case *appsv1.Deployment:
+		var replicas int32 = 1
+		if obj.Spec.Replicas != nil {
+			replicas = *obj.Spec.Replicas
+		}
+		return &autoscalingv1.Scale{
+			ObjectMeta: obj.ObjectMeta,
```
Comment (alvaroaleman) [inline pkg/client/fake/client.go:1334]: I think its only namespace, name, uid, resourceVersion and creationTimestamp we should copy over, not for example labels or annotations. Please update this and the corresponding tests.

File: pkg/client/fake/client_test.go [hunk 9]
```diff
@@ -2068,6 +2070,70 @@ var _ = Describe("Fake client", func() {
...
+		obj := &corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "foo",
+			},
+		}
+		cl := NewClientBuilder().WithScheme(scheme).WithObjects(obj).Build()
+
+		scale := &autoscalingv1.Scale{Spec: autoscalingv1.ScaleSpec{Replicas: 2}}
+		Expect(cl.SubResource(subResourceScale).Get(context.Background(), obj, scale)).ToNot(Succeed())
+		Expect(cl.SubResource(subResourceScale).Update(context.Background(), obj, client.WithSubResourceBody(scale))).ToNot(Succeed())
```
Comment (alvaroaleman) [inline pkg/client/fake/client_test.go:2088]: Could we match the error rather than just checking for an error? Right now, this test would for example incorrectly pass if you did a GET for an object that doesn't exist

File: pkg/client/fake/client_test.go [hunk 10]
```diff
@@ -2068,6 +2070,70 @@ var _ = Describe("Fake client", func() {
...
+		objOriginal := obj.DeepCopy()
+
+		scale := &autoscalingv1.Scale{}
+		Expect(cl.SubResource(subResourceScale).Get(context.Background(), obj, scale)).NotTo(HaveOccurred())
+
+		actual := &appsv1.Deployment{ObjectMeta: metav1.ObjectMeta{Name: obj.Name}}
+		Expect(cl.Get(context.Background(), client.ObjectKeyFromObject(actual), actual)).To(Succeed())
+
+		objOriginal.APIVersion = scale.APIVersion
+		objOriginal.Kind = scale.Kind
```
Comment (alvaroaleman) [inline pkg/client/fake/client_test.go:2110]: These lines confuse me. Neither `objOriginal` nor `scale` should have either `APIVersion` or `Kind` set - Why is this needed? Same question for the update test.

File: pkg/client/fake/client_test.go [hunk 11]
```diff
@@ -2068,6 +2070,70 @@ var _ = Describe("Fake client", func() {
...
+		}
+		cl := NewClientBuilder().WithScheme(scheme).WithObjects(obj).Build()
+
+		scale := &autoscalingv1.Scale{Spec: autoscalingv1.ScaleSpec{Replicas: 2}}
+		Expect(cl.SubResource(subResourceScale).Get(context.Background(), obj, scale)).ToNot(Succeed())
+		Expect(cl.SubResource(subResourceScale).Update(context.Background(), obj, client.WithSubResourceBody(scale))).ToNot(Succeed())
+	})
+
+	It("should be able to Get scale subresources", func() {
+		obj := &appsv1.Deployment{
```
Comment (alvaroaleman) [inline pkg/client/fake/client_test.go:2092]: It would be good to test this for all object kinds where we support this, maybe you can loop over a list of deployment/replicaset/statefulset/daemonset for that?

File: pkg/client/fake/client_test.go [hunk 12]
```diff
@@ -2068,6 +2070,70 @@ var _ = Describe("Fake client", func() {
...
+		scale := &autoscalingv1.Scale{}
+		Expect(cl.SubResource(subResourceScale).Get(context.Background(), obj, scale)).NotTo(HaveOccurred())
+
+		actual := &appsv1.Deployment{ObjectMeta: metav1.ObjectMeta{Name: obj.Name}}
+		Expect(cl.Get(context.Background(), client.ObjectKeyFromObject(actual), actual)).To(Succeed())
+
+		objOriginal.APIVersion = scale.APIVersion
+		objOriginal.Kind = scale.Kind
+		objOriginal.ResourceVersion = scale.ResourceVersion
+		objOriginal.Spec.Replicas = ptr.To(scale.Spec.Replicas)
```
Comment (alvaroaleman) [inline pkg/client/fake/client_test.go:2112]: This test needs to verify that they are equal, right now it doesn't do that

**Comment by alvaroaleman** [comment]
@TheSpiritXIII will you be able to rebase and resolve the remaining comments?

**Comment by TheSpiritXIII** [comment]
@alvaroaleman I rebased and addressed all comments. Please let me know if there's anything else! Thanks.

**Review by alvaroaleman — APPROVED** [review]
Thank you!
